mewarpx/mewarpx/diags_store/field_diagnostic.py

```diff
@@ -20,121 +21,259 @@
-        if mwxrun.get_it() == mwxrun.simulation.max_steps:
-            self.do_post_processing()
+        logger.info("Analyzing fields...")
+        self.it = mwxrun.get_it()
+
+        if self.process_phi:
+            data = mwxrun.get_gathered_phi_grid(include_ghosts=False)
+            self.process_field(
+                data=data,
+                titlestr='Electrostatic potential',
+                plottype='phi', draw_image=True, default_ticks=True,
+                draw_contourlines=False)
+            # Optionally generate barrier index plot if requested
+            if self.plot and (self.barrier_slices is not None):
+                self.plot_barrier_slices(data, self.barrier_slices)
+        if self.process_E:
+            raise NotImplementedError("E-field processing not yet implemented.")
+            self.process_field(
+                data=None,
+                titlestr='Electric field strength',
+                plottype='E', draw_image=True, default_ticks=True,
+                draw_contourlines=False)
+        if self.process_rho:
+            # assume that rho_fp still holds the net charge density
+            data = (
+                mwxrun.get_gathered_rho_grid(include_ghosts=False)
+                / mwxconstants.e * 1e-6
+            )[:,:,0]
+            self.process_field(
+                data=data,
+                titlestr='Net charge density',
+                plottype='rho', draw_image=True, default_ticks=True,
+                draw_contourlines=False)
+
+            # deposit the charge density for each species
+            for species in self.species_list:
+                data = (
+                    mwxrun.get_gathered_rho_grid(
+                        species_name=species.name, include_ghosts=False
+                    ) / mwxconstants.e * 1e-6
+                )[:,:,0]
+                self.process_field(
+                    data=data,
+                    titlestr=f'{species.name} charge density',
+                    plottype='rho', draw_image=True, default_ticks=True,
+                    draw_contourlines=False)
+
+        logger.info("Finished analyzing fields")
+
+    def process_field(self, data, titlestr, plottype=None,
+                      **kwargs):
+        """Save given field to file, and optionally plot it as well.
+
+        Other kwargs are passed on to plotting.
+
+        Arguments:
+            data (numpy.ndarray): Array to output
+            titlestr (string): String to use as first part of the file name.
+                Also used for the title in plotting.
+            plottype: Choose 'phi', 'E', or 'rho' to set correct labels
+        """
+
+        if mwxrun.me == 0:
+            try:
+                fileprefix = self.get_fileprefix(titlestr)
+                # Save full field
+                np.save(fileprefix + '.npy', data)
+            except Exception as exc:
+                warnings.warn("Saving {} failed with error {}".format(
+                    titlestr, exc))
```

Comment:
I think we previously agreed to use fstrings moving forward to stay more modern and "pythonic".

```suggestion
                warnings.warn(f"Saving {titlestr} failed with error {exc}")
```